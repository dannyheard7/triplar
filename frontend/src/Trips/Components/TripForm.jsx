import React from "react";

import FormFieldGroup from "Forms/Components/FormFieldGroup";
import NonFieldErrors from "Forms/Components/NonFieldErrors";
import CitySearchFieldContainer from "Forms/Containers/CitySearchFieldContainer";


export default class TripForm extends React.Component {
    render () {
        // TODO: convert to map to allow multiple locations and remove this mess
        let location = (typeof this.props.trip.locations !== 'undefined') ? this.props.trip.locations[0] : "";

        return(
            <form onSubmit={ this.props.onSubmit }>
                <NonFieldErrors errors={this.props.errors.non_field_errors} />
                <CitySearchFieldContainer  errors={this.props.errors.locations}
                                        onChange={this.props.onChange} value={location} />
                <FormFieldGroup errors={this.props.errors.name} label="Trip Name" name="name" type="text"
                                onChange={this.props.onChange} value={this.props.trip.name} required="true"/>
                <FormFieldGroup errors={this.props.errors.start_date} label="Start Date" name="start_date" type="date"
                                onChange={this.props.onChange} value={this.props.trip.start_date} required="true" />
                <FormFieldGroup errors={this.props.errors.end_date} label="End Date" name="end_date" type="date"
                                onChange={this.props.onChange} value={this.props.trip.end_date} required="true" />
                <button type="submit" className="btn btn-primary" >{this.props.submitLabel}</button>
            </form>
        );
    }
}

TripForm.defaultProps = {
    trip: {}
};
