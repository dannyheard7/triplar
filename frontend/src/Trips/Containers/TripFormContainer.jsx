import React from "react";
import TripForm from "Trips/Components/TripForm";

export default class TripFormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            trip: this.props.trip,
            errors: []
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }


    handleChange(event) {
        let trip = this.state.trip;
        trip[event.target.name] = event.target.value;

        this.setState({trip: trip});
    }

    handleSubmit(e) {
        e.preventDefault();

        this.props.apiFunction(this.state.trip).then(response => {
            if(response.status === 400) {
                this.setState({errors: response.data})
            } else {
                this.setState({trip: {}, errors: []});
                this.props.onSuccess(response.data);
            }
        });
    }

    render () {
        return (
            <TripForm onChange={this.handleChange} onSubmit={this.handleSubmit} submitLabel={this.props.submitLabel}
                      errors={this.state.errors} suggestions={this.state.suggestions} trip={this.state.trip} />
        )
    }

}

TripFormContainer.defaultProps = {
    trip: {}
};