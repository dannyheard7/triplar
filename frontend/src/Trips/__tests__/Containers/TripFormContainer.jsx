import React from "react";
import {shallow} from "enzyme";
import ShallowRenderer from 'react-test-renderer/shallow';
import TripForm from "Trips/Components/TripForm";
import TripFormContainer from "Trips/Containers/TripFormContainer";

const renderer = new ShallowRenderer();

const faker = require('faker');

describe('<TripFormContainer />', () => {
    const event = {
        preventDefault: () => {
        },
        target: {
            value: "",
            name: "",
        },
    };

    const onSuccess = jest.fn();

    const apiFunc = jest.fn().mockImplementation(trip => {
        return new Promise((resolve, reject) => {
            let response = {status: 200, data: trip};
            resolve(response);
        });
    });

    const props = {apiFunction: apiFunc, onSuccess: onSuccess};

    afterEach(() => {
        apiFunc.mockClear();
        onSuccess.mockClear();
    });

    test('renders correctly', () => {
        const result = renderer.render(<TripFormContainer {...props} />);
        expect(result).toMatchSnapshot();
    });

    test('calls handle change after receiving change from <TripForm />', (function () {
        const spy = jest.spyOn(TripFormContainer.prototype, 'handleChange');

        const container = shallow(<TripFormContainer {...props} />);
        container.find(TripForm).prop('onChange')(event);

        expect(spy).toBeCalled();
    }));

    test('calls handle submit after receiving submit from <TripForm />', async() => {
        const spy = jest.spyOn(TripFormContainer.prototype, 'handleSubmit');

        let container = shallow(<TripFormContainer {...props}  />);

        container.find(TripForm).prop('onSubmit')(event);
        await Promise.resolve();

        expect(spy).toBeCalled();
    });

    test('calls apiFunction prop during handleSubmit()', async() => {
        let container = shallow(<TripFormContainer {...props}  />);

        container.instance().handleSubmit(event);
        await Promise.resolve();

        expect(apiFunc).toBeCalled();
    });

    test('calls onSuccess prop on api call success', async() => {
        let container = shallow(<TripFormContainer {...props}  />);

        container.find(TripForm).prop('onSubmit')(event);
        await Promise.resolve();

        expect(onSuccess).toBeCalled();
    });

    test('updates the state when handleChange is called', () => {
        const container = shallow(<TripFormContainer {...props} />);
        const instance = container.instance();
        const name = faker.random.word();

        instance.handleChange({'target': {'name': 'name', 'value': name}});
        expect(container.state('trip').name).toEqual(name);
    });

    test('errors state change updates TripForm Prop', () => {
        const container = shallow(<TripFormContainer {...props} />);
        expect(container.find(TripForm).props().errors).toEqual(container.state('errors'));

        let errors = jest.fn();
        container.setState({errors: errors});

        expect(container.find(TripForm).props().errors).toEqual(errors);
    });

    test('unsuccesful api call updates state errors', async () => {
        const spy = jest.fn();
        let trip = {'id': faker.random.number(), 'name': faker.random.word()};
        const errors = [faker.random.word(), faker.random.word()];
        const apiFunc = jest.fn().mockReturnValueOnce(new Promise((resolve, reject) => {
            let response = {status: 400, data: errors};
            resolve(response);
        }));

        const container = shallow(<TripFormContainer apiFunction={apiFunc} onSuccess={onSuccess} trip={trip}/>);
        container.instance().handleSubmit(event);
        await Promise.resolve();

        expect(spy).not.toBeCalled();
        expect(container.state('errors')).toEqual(errors);
    });

});