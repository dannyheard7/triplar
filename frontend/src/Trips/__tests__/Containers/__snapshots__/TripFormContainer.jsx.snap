// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripFormContainer /> renders correctly 1`] = `
<TripForm
  errors={Array []}
  onChange={[Function]}
  onSubmit={[Function]}
  submitLabel={undefined}
  suggestions={undefined}
  trip={Object {}}
/>
`;
